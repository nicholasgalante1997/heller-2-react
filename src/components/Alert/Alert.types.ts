export interface AlertProps {
  severity: Severity;
  message: {
    heading: string;
    body: string;
  };
  dismissable?: boolean;
  onDismiss?: () => void;
}

type Severity = 'info' | 'warning' | 'error' | 'success';
