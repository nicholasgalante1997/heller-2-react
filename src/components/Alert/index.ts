export { Alert } from './Alert';
export { AlertsCtxProvider, useAlertsCtx } from './AlertsCtx';
export { type AlertProps } from './Alert.types';