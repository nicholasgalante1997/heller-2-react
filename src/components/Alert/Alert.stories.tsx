import React from 'react';
import { type Meta, type StoryObj } from '@storybook/react';
import { Alert } from './Alert';
import { AlertsCtxProvider, useAlertsCtx } from './AlertsCtx';
import { Button } from '../Button';

const meta: Meta<typeof Alert> = {
  title: 'REACT/Components/Alert',
  component: Alert
};

export default meta;

export const Info: StoryObj<typeof Alert> = {
  args: {
    dismissable: false,
    severity: 'info',
    message: {
      heading: 'Join our Discord!',
      body: 'Visit the link attached here to join our discord community. You can receive access to early release tech merchandise.'
    }
  },
  render: (args) => <Alert {...args} />
};

export const Warning: StoryObj<typeof Alert> = {
  args: {
    dismissable: false,
    severity: 'warning',
    message: {
      heading: 'Warning! Beta Feature',
      body: 'This feature is experimental and subject to change with further updates.'
    }
  },
  render: (args) => <Alert {...args} />
};

export const Error: StoryObj<typeof Alert> = {
  args: {
    dismissable: false,
    severity: 'error',
    message: {
      heading: 'Unable To Sign Up',
      body: 'A user with this email address already exists.'
    }
  },
  render: (args) => <Alert {...args} />
};

export const Success: StoryObj<typeof Alert> = {
  args: {
    dismissable: false,
    severity: 'success',
    message: {
      heading: 'Nice! Story Submitted',
      body: 'Check your email for a confirmation that we have received your submission.'
    }
  },
  render: (args) => <Alert {...args} />
};

// Alerts Context Example

function AlertsCtxExample() {
  return (
    <AlertsCtxProvider>
      <AlertsCtxButton />
    </AlertsCtxProvider>
  );
}

function AlertsCtxButton() {
  const { dispatchAlert } = useAlertsCtx();
  function onClick() {
    dispatchAlert({
      message: {
        heading: 'Critical Error',
        body: `You have ${Math.ceil(Math.random() * 60)} seconds to disengage.`
      },
      severity: 'error',
      ttd: 300000,
      dismissable: true
    });
  }
  return (
    <Button style={{ position: 'absolute', bottom: 24 }} onClick={onClick}>
      Dispatch New Alert
    </Button>
  );
}

export const AlertsCtxMain: StoryObj = {
  render: () => <AlertsCtxExample />
};
