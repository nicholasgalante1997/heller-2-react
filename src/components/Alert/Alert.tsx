import React, { memo } from 'react';
import classNames from 'classnames';
import { type AlertProps } from './Alert.types';

import { Body } from '../Body';
import { Heading } from '../Heading';
import { Conditional } from '../_internal/Conditional';

const alertClassNames = {
  root: 'h-alert',
  body: 'h-alert-body-content',
  icon: 'h-alert-dismiss-icon-container'
} as const;

function AlertComponent(props: AlertProps) {
  const { message, severity, dismissable = false, onDismiss, className, id, ...rest } = props;
  return (
    <div {...rest} id={id} className={classNames(alertClassNames.root, className)} role="alert">
      <span id="h-alert-severity-indicator" data-severity={severity} />
      <div data-shrink={dismissable ? 'shrink' : 'none'} className={alertClassNames.body}>
        <Heading as="h6" data-severity={severity}>
          {message.heading}
        </Heading>
        <Body as="p" accent bold>
          {message.body}
        </Body>
      </div>
      <Conditional condition={dismissable}>
        <span className={alertClassNames.icon} role="button" onClick={onDismiss}>
          <svg width="40px" height="40px" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
            <g>
              <path
                d="M16 16L12 12M12 12L8 8M12 12L16 8M12 12L8 16"
                stroke="#000000"
                strokeWidth="2"
                strokeLinecap="round"
                strokeLinejoin="round"
              />
            </g>
          </svg>
        </span>
      </Conditional>
    </div>
  );
}

export const Alert = memo(AlertComponent);
