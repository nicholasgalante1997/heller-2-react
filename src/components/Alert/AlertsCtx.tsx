import React, { createContext, useContext, useState, useEffect, useCallback } from 'react';
import isEqual from 'lodash.isequal';
import { Alert } from './Alert';
import { type AlertProps } from './Alert.types';

type AggregateAlertProps = { ttd?: number } & AlertProps;

interface AlertsCtxType {
  dispatchAlert: (alertConfiguration: AggregateAlertProps) => void;
}

const alertsContextDefaultState: AlertsCtxType = {
  dispatchAlert(alertConfiguration) {
    /** null implementation */
  }
};

const AlertsContext = createContext<AlertsCtxType>(alertsContextDefaultState);
const useAlertsCtx = () => useContext(AlertsContext);

interface AlertsCtxProviderProps {
  children: React.ReactNode | React.ReactNode[];
}

const alertsContextClassNames = {
  root: 'h-alerts-context-provider-root-mount'
} as const;

function AlertsCtxProvider({ children }: AlertsCtxProviderProps) {
  const [alerts, setAlerts] = useState<AggregateAlertProps[]>([]);

  useEffect(
    function () {
      if (alerts.length > 0) {
        const alert = alerts[0];
        const ttd = alert?.ttd || 7000;
        const timeout = setTimeout(function () {
          removeAlert(alert);
        }, ttd);
        return () => {
          clearTimeout(timeout);
        };
      }
    },
    [alerts]
  );

  const dispatchAlert = useCallback(
    (alertProps: AggregateAlertProps) => {
      let alertExists = false;
      for (const alertConfig of alerts) {
        if (isEqual(alertConfig, alertProps)) alertExists = true;
      }
      if (!alertExists) setAlerts((current) => [...current, alertProps]);
    },
    [alerts, setAlerts]
  );

  const removeAlert = useCallback(
    (alertConfig: AggregateAlertProps) => {
      const alert = alerts.find((existingAlertConfig) => isEqual(existingAlertConfig, alertConfig));
      if (alert) {
        setAlerts((existingAlerts) =>
          existingAlerts.filter((existingAlertConfig) => !isEqual(existingAlertConfig, alert))
        );
      }
    },
    [alerts, setAlerts]
  );

  const sortAlertProps = (alertConfig: AggregateAlertProps) => {
    const { message, severity, className, dismissable, id, onDismiss, role } = alertConfig;
    const props = {};
    Object.assign(props, {
      message,
      severity,
      className,
      id,
      dismissable,
      role
    });
    if (dismissable) {
      function aggregateOnDismiss() {
        removeAlert(alertConfig);
        if (typeof onDismiss !== 'undefined') {
          onDismiss();
        }
      }
      Object.assign(props, { onDismiss: aggregateOnDismiss });
    }
    return props as AlertProps;
  };

  return (
    <AlertsContext.Provider value={{ dispatchAlert }}>
      <span className={alertsContextClassNames.root}>
        {alerts.map((alertConfig) => (
          <Alert {...sortAlertProps(alertConfig)} />
        ))}
      </span>
      {children}
    </AlertsContext.Provider>
  );
}

export { AlertsCtxProvider, useAlertsCtx };
