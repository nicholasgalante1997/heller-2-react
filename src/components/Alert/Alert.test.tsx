import React from 'react';
import { render } from '@testing-library/react';
import { Alert } from './Alert';

describe('Alert', () => {
  test('Snapshot', () => {
    const body = 'A longer description of the event or cause of the alert dispatch.';
    const heading = 'A brief note';
    for (const severity of ['info', 'error', 'success', 'warning']) {
      const { container } = render(<Alert severity={severity as any} message={{ heading, body }} />);
      expect(container).toMatchSnapshot();
    }
  });
});
