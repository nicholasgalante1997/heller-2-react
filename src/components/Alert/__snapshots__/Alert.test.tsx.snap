// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Snapshot 1`] = `
<div>
  <div
    class="h-alert"
    role="alert"
  >
    <span
      data-severity="info"
      id="h-alert-severity-indicator"
    />
    <div
      class="h-alert-body-content"
      data-shrink="none"
    >
      <h6
        class="h-heading-600"
        data-severity="info"
      >
        A brief note
      </h6>
      <p
        class="h-body h-body-bold h-body-accent"
      >
        A longer description of the event or cause of the alert dispatch.
      </p>
    </div>
  </div>
</div>
`;

exports[`Alert Snapshot 2`] = `
<div>
  <div
    class="h-alert"
    role="alert"
  >
    <span
      data-severity="error"
      id="h-alert-severity-indicator"
    />
    <div
      class="h-alert-body-content"
      data-shrink="none"
    >
      <h6
        class="h-heading-600"
        data-severity="error"
      >
        A brief note
      </h6>
      <p
        class="h-body h-body-bold h-body-accent"
      >
        A longer description of the event or cause of the alert dispatch.
      </p>
    </div>
  </div>
</div>
`;

exports[`Alert Snapshot 3`] = `
<div>
  <div
    class="h-alert"
    role="alert"
  >
    <span
      data-severity="success"
      id="h-alert-severity-indicator"
    />
    <div
      class="h-alert-body-content"
      data-shrink="none"
    >
      <h6
        class="h-heading-600"
        data-severity="success"
      >
        A brief note
      </h6>
      <p
        class="h-body h-body-bold h-body-accent"
      >
        A longer description of the event or cause of the alert dispatch.
      </p>
    </div>
  </div>
</div>
`;

exports[`Alert Snapshot 4`] = `
<div>
  <div
    class="h-alert"
    role="alert"
  >
    <span
      data-severity="warning"
      id="h-alert-severity-indicator"
    />
    <div
      class="h-alert-body-content"
      data-shrink="none"
    >
      <h6
        class="h-heading-600"
        data-severity="warning"
      >
        A brief note
      </h6>
      <p
        class="h-body h-body-bold h-body-accent"
      >
        A longer description of the event or cause of the alert dispatch.
      </p>
    </div>
  </div>
</div>
`;
